modernize.py

```diff
@@ -18,92 +18,319 @@
+    "libmodernize.fixes.fix_file",
+    "libmodernize.fixes.fix_filter",
+    "libmodernize.fixes.fix_import",
+    "libmodernize.fixes.fix_imports_six",
+    "libmodernize.fixes.fix_input_six",
+    "libmodernize.fixes.fix_int_long_tuple",
+    "libmodernize.fixes.fix_itertools_imports_six",
+    "libmodernize.fixes.fix_itertools_six",
+    "libmodernize.fixes.fix_map",
+    "libmodernize.fixes.fix_metaclass",
+    "libmodernize.fixes.fix_next",
+    "libmodernize.fixes.fix_print",
+    "libmodernize.fixes.fix_raise",
+    "libmodernize.fixes.fix_raise_six",
+    "libmodernize.fixes.fix_unichr",
+    "libmodernize.fixes.fix_unicode_type",
+    "libmodernize.fixes.fix_urllib_six",
+    "libmodernize.fixes.fix_xrange_six",
+    "libmodernize.fixes.fix_zip",
+}
+
+
+def modernize_string(source, disabled_fixers=None):
+    """
+    Apply "modernize" to string with python source code.
+    :param source: source code.
+    :param disabled_fixers: fixers which are not going to be used in this file processing.
+    :return: "Modernized" source code
+    :type source six.text_type
+    :type disabled_fixers set
+    :rtype six.text_type
+    """
+    if disabled_fixers is None:
+        disabled_fixers = set()
+
+    fixers = FIXERS - set(disabled_fixers)
+    tool = refactor.RefactoringTool(sorted(fixers))
+    tree = tool.refactor_string(source, "<stdin>")
+    new_source = six.text_type(tree)
+
+    return new_source
+
+
+def get_diff(new_source, original_source, file_path):
+    """
+    Get diff from two strings.
+    :param new_source:
+    :param original_source:
+    :param file_path:
+    :return: string with diff
+    :rtype six.text_type
+    """
+    diff = difflib.unified_diff(
+        original_source.splitlines(),
+        new_source.splitlines(),
+        file_path,
+        file_path,
+        "(original)",
+        "(refactored)",
+        lineterm="",
+    )
+
+    diff_text = "\n".join(list(diff))
 
-from scalyr_agent.__scalyr__ import get_package_root
+    return diff_text
 
 
-def modernize(files, params_str=""):
+def parse_todos(source):
     """
-    :param files: list of files paths.
-    :param params_str:  additional parameters to python-modernize
+    :param source: string with python source code
+    :type source six.text_type
     :return:
     """
-    process = subprocess.Popen(
-        "python-modernize {} -w -n {}".format(params_str, " ".join(list(files))),
-        shell=True,
-        stderr=subprocess.PIPE,
-        stdout=subprocess.PIPE
+    # pattern to find start of the area [start of 2->TOD0]
+    start_pattern = r"(?!\n)[\s]*#\s*\[\s*start of 2->TODO\s*[a-zA-Z._ ]*\s*\]"
+    # pattern to find end of the area [start of 2->TOD0]
+    end_pattern = r"#\s*\[\s*end of 2->TOD[O0]\s*\]"
+    todo_area_pattern = re.compile(
+        r"{start}((?:(?!{end}).)*){end}".format(start=start_pattern, end=end_pattern),
+        re.DOTALL,
     )
+    found = todo_area_pattern.findall(source)
```

Comment:
Let's do a safety check here.. I want to make sure someone did not forget to add and `end of 2->TODO`.  So, for each found TODO area, check to make sure (1) there's no `start of 2->TODO` in there and that the area itself is not more than 500 characters (500 is a bit arbitrary... but any number that's reasonable is fine)

Reply:
I have already added some areas which bigger than 500 characters, so I think it will be ok to set a limit to 1000.